File: alumnium/drivers/base_driver.py
from abc import ABC, abstractmethod
from contextlib import contextmanager
from typing import Iterator

from alumnium.accessibility import BaseAccessibilityTree

from .keys import Key


class BaseDriver(ABC):
    _accessibility_tree: BaseAccessibilityTree

    @contextmanager
    def capture_accessibility_tree(self, tree: BaseAccessibilityTree = None) -> Iterator[BaseAccessibilityTree]:
        if tree:
            self._accessibility_tree = tree
        else:
            self._accessibility_tree = self._fetch_accessibility_tree()
        try:
            yield self._accessibility_tree
        finally:
            self._accessibility_tree = None

    @abstractmethod
    def click(self, id: int):
        pass

    @abstractmethod
    def drag_and_drop(self, from_id: int, to_id: int):
        pass

    @abstractmethod
    def press_key(self, key: Key):
        pass

    @abstractmethod
    def quit(self):
        pass

    @abstractmethod
    def back(self):
        pass

    @property
    @abstractmethod
    def screenshot(self) -> str:
        pass

    @abstractmethod
    def select(self, id: int, option: str):
        pass

    @property
    @abstractmethod
    def title(self) -> str:
        pass

    @abstractmethod
    def type(self, id: int, text: str):
        pass

    @property
    @abstractmethod
    def url(self) -> str:
        pass

    @abstractmethod
    def _fetch_accessibility_tree(self) -> BaseAccessibilityTree:
        pass

File: ts/src/drivers/BaseDriver.ts
import { RawAccessibilityTree } from '../accessibility/RawAccessibilityTree.js';
import { Key } from './keys.js';

export abstract class BaseDriver {
  abstract get accessibilityTree(): RawAccessibilityTree;
  abstract getAccessibilityTree(): Promise<RawAccessibilityTree>;
  abstract click(id: number): void | Promise<void>;
  abstract dragAndDrop(fromId: number, toId: number): void | Promise<void>;
  abstract pressKey(key: Key): void | Promise<void>;
  abstract quit(): void | Promise<void>;
  abstract back(): void | Promise<void>;
  abstract screenshot(): string | Promise<string>;
  abstract select(id: number, option: string): void | Promise<void>;
  abstract title(): string | Promise<string>;
  abstract type(id: number, text: string): void | Promise<void>;
  abstract url(): string | Promise<string>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-redundant-type-constituents
  abstract findElement(id: number): any | Promise<any>;
  abstract hover(id: number): void | Promise<void>;
}
